File: apps/web/src/api/sessions.ts
import { requestJson } from "./client";
import type {
  DeleteResult,
  Message,
  PageResult,
  Run,
  RunCreate,
  RunCreateAccepted,
  Session,
  SessionCreate,
  SessionUpdate,
} from "./types";

function pageQuery(params: Record<string, string | number | undefined>): string {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === "") continue;
    search.set(key, String(value));
  }
  const encoded = search.toString();
  return encoded ? `?${encoded}` : "";
}

export function listSessions(
  page = 1,
  pageSize = 20,
  q?: string,
): Promise<PageResult<Session>> {
  return requestJson<PageResult<Session>>(
    `/sessions${pageQuery({ page, page_size: pageSize, q })}`,
  );
}

export function createSession(payload: SessionCreate): Promise<Session> {
  return requestJson<Session>("/sessions", { method: "POST", body: JSON.stringify(payload) });
}

export function getSession(sessionId: string): Promise<Session> {
  return requestJson<Session>(`/sessions/${encodeURIComponent(sessionId)}`);
}

export function updateSession(sessionId: string, payload: SessionUpdate): Promise<Session> {
  return requestJson<Session>(`/sessions/${encodeURIComponent(sessionId)}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteSession(sessionId: string): Promise<DeleteResult> {
  return requestJson<DeleteResult>(`/sessions/${encodeURIComponent(sessionId)}`, { method: "DELETE" });
}

export function listSessionMessages(
  sessionId: string,
  pageSize = 20,
  beforePosition?: number,
): Promise<PageResult<Message>> {
  return requestJson<PageResult<Message>>(
    `/sessions/${encodeURIComponent(sessionId)}/messages${pageQuery({
      page_size: pageSize,
      before_position: beforePosition,
    })}`,
  );
}

export function createRun(sessionId: string, payload: RunCreate): Promise<RunCreateAccepted> {
  return requestJson<RunCreateAccepted>(`/sessions/${encodeURIComponent(sessionId)}/runs`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function listSessionRuns(
  sessionId: string,
  page = 1,
  pageSize = 20,
  q?: string,
): Promise<PageResult<Run>> {
  return requestJson<PageResult<Run>>(
    `/sessions/${encodeURIComponent(sessionId)}/runs${pageQuery({
      page,
      page_size: pageSize,
      q,
    })}`,
  );
}

File: apps/api/server/routes/sessions.py
from __future__ import annotations

from typing import Annotated

from application.sessions import SessionService
from contracts.api import ApiEnvelope, DeleteResult, Page, PageResult
from contracts.runs import RunCreate, RunCreateAccepted, RunRead
from contracts.sessions import MessageRead, SessionCreate, SessionRead, SessionUpdate
from fastapi import APIRouter, Depends, Query, Request
from runtime.run_service import RunService

from server.dependencies import get_run_service, get_session_service
from server.responses import success_response

router = APIRouter(prefix="/sessions", tags=["sessions"])
SessionServiceDep = Annotated[SessionService, Depends(get_session_service)]
RunServiceDep = Annotated[RunService, Depends(get_run_service)]


@router.get("", response_model=ApiEnvelope[PageResult[SessionRead]])
async def list_sessions(
    request: Request,
    service: SessionServiceDep,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    q: str | None = Query(default=None, max_length=200),
):
    """分页返回会话列表，可选按标题子串过滤。"""

    return success_response(
        request,
        await service.list(Page(page=page, page_size=page_size), q=q),
    )


@router.post("", status_code=201, response_model=ApiEnvelope[SessionRead])
async def create_session(
    request: Request,
    payload: SessionCreate,
    service: SessionServiceDep,
):
    """创建一个新的会话。"""

    return success_response(request, await service.create(payload), status_code=201)


@router.get("/{session_id}", response_model=ApiEnvelope[SessionRead])
async def get_session(
    request: Request,
    session_id: str,
    service: SessionServiceDep,
):
    """读取指定会话。"""

    return success_response(request, await service.get(session_id))


@router.post("/{session_id}/runs", status_code=202, response_model=ApiEnvelope[RunCreateAccepted])
async def create_run(
    request: Request,
    session_id: str,
    payload: RunCreate,
    service: RunServiceDep,
):
    """按 Session 当前数据源创建一个后台分析 Run。"""

    return success_response(request, await service.create(session_id, payload), status_code=202)


@router.get("/{session_id}/runs", response_model=ApiEnvelope[PageResult[RunRead]])
async def list_session_runs(
    request: Request,
    session_id: str,
    service: RunServiceDep,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    q: str | None = Query(default=None, max_length=2000),
):
    """只读返回会话内已保存的 Run 历史，不触发重放。"""

    return success_response(
        request,
        await service.list_for_session(
            session_id,
            Page(page=page, page_size=page_size),
            q=q,
        ),
    )


@router.patch("/{session_id}", response_model=ApiEnvelope[SessionRead])
async def update_session(
    request: Request,
    session_id: str,
    payload: SessionUpdate,
    service: SessionServiceDep,
):
    """更新指定会话的基础信息。"""

    return success_response(request, await service.update(session_id, payload))


@router.delete("/{session_id}", response_model=ApiEnvelope[DeleteResult])
async def delete_session(
    request: Request,
    session_id: str,
    service: SessionServiceDep,
):
    """删除指定会话。"""

    await service.delete(session_id)
    return success_response(request, {"deleted": True})


@router.get("/{session_id}/messages", response_model=ApiEnvelope[PageResult[MessageRead]])
async def list_session_messages(
    request: Request,
    session_id: str,
    service: SessionServiceDep,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    before_position: int | None = Query(default=None, ge=1),
):
    """返回最近一窗已保存消息；before_position 用于向上加载更早消息。"""

    del page
    return success_response(
        request,
        await service.list_messages(
            session_id,
            Page(page=1, page_size=page_size),
            before_position=before_position,
        ),
    )
